File: packages/sdk/src/commands/withdraw.ts
// withdraw command — WD-9 (SDK side).

import chalk from "../lib/chalk-shim.js";
import { resolveRegistryUrl, RegistryError } from "../lib/registry.js";

export interface WithdrawOptions {
  dropId: string;
  registryUrl?: string;
  bearerToken?: string;
  confirm?: boolean;
}

export async function run(opts: WithdrawOptions): Promise<void> {
  try {
    if (!opts.confirm) {
      process.stderr.write(
        `Re-run with --confirm to withdraw '${opts.dropId}'. Bundles will stay on R2 ` +
          `so already-installed users keep working, but the drop will be hidden from ` +
          `browse + trending.\n`,
      );
      process.exit(1);
    }

    const token = opts.bearerToken ?? process.env.WINDY_REGISTRY_TOKEN;
    if (!token) {
      throw new Error("no Bearer token. Pass --token <jwt> or set $WINDY_REGISTRY_TOKEN");
    }

    const url = `${resolveRegistryUrl(opts)}/api/v1/drops/${encodeURIComponent(opts.dropId)}`;
    const r = await fetch(url, {
      method: "DELETE",
      headers: { authorization: `Bearer ${token}` },
    });
    if (!r.ok) {
      let body;
      try { body = await r.json(); } catch { body = await r.text(); }
      throw new RegistryError(`withdraw failed: ${r.status}`, r.status, body);
    }

    process.stdout.write(`${chalk.green("✓")} withdrew ${opts.dropId}\n`);
    process.stdout.write(
      `  bundle remains on R2; installed users keep working.\n`,
    );
    process.exit(0);
  } catch (e) {
    process.stderr.write(`${chalk.red("✗")} ${(e as Error).message}\n`);
    const msg = (e as Error).message;
    if (msg.includes("401") || msg.includes("Bearer")) process.exit(2);
    if (msg.includes("403")) process.exit(2);
    if (msg.includes("404")) process.exit(4);
    process.exit(1);
  }
}

File: python/sdk/src/windy_drops/commands/withdraw.py
"""withdraw command — WD-9 (SDK side)."""

from __future__ import annotations

import os
import sys

import httpx
import typer

from ..lib.registry import resolve_registry_url


def run(
    drop_id: str = typer.Argument(..., help="Drop id to withdraw"),
    registry_url: str | None = typer.Option(None, "--registry-url"),
    token: str | None = typer.Option(None, "--token"),
    confirm: bool = typer.Option(False, "--confirm"),
) -> None:
    """Hide a drop from search; existing installs keep working."""
    if not confirm:
        typer.echo(
            f"Re-run with --confirm to withdraw '{drop_id}'. Bundles will stay on R2 "
            "so already-installed users keep working, but the drop will be hidden "
            "from browse + trending.",
            err=True,
        )
        raise typer.Exit(code=1)

    bearer = token or os.environ.get("WINDY_REGISTRY_TOKEN")
    if not bearer:
        typer.echo("✗ no Bearer token. Pass --token <jwt> or set $WINDY_REGISTRY_TOKEN", err=True)
        raise typer.Exit(code=2)

    url = f"{resolve_registry_url(registry_url=registry_url)}/api/v1/drops/{drop_id}"
    r = httpx.delete(url, headers={"authorization": f"Bearer {bearer}"}, timeout=30.0)
    if r.status_code == 204:
        typer.echo(f"✓ withdrew {drop_id}")
        typer.echo("  bundle remains on R2; installed users keep working.")
        return
    if r.status_code in (401, 403):
        typer.echo(f"✗ withdraw failed: {r.status_code}", err=True)
        raise typer.Exit(code=2)
    if r.status_code == 404:
        typer.echo("✗ drop_not_found", err=True)
        raise typer.Exit(code=4)
    typer.echo(f"✗ withdraw failed: {r.status_code} {r.text}", err=True)
    raise typer.Exit(code=1)
